piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+from contextlib import asynccontextmanager as acm
+from datetime import datetime
+import time
+import math
+import base64
+import hmac
+import hashlib
+import wsproto
+from uuid import uuid4
+
+import asks
+import tractor
+import trio
+from trio_typing import TaskStatus
+from fuzzywuzzy import process as fuzzy
+import pendulum
+import numpy as np
+
+from piker._cacheables import open_cached_client
+from piker.log import get_logger
+from ._util import DataUnavailable
+from piker.pp import config
+from ..data.types import Struct
+from ..data._web_bs import (
+    open_autorecon_ws,
+    NoBsWs,
+)
+
+log = get_logger(__name__)
+
+_ohlc_dtype = [
+    ('index', int),
+    ('time', int),
+    ('open', float),
+    ('high', float),
+    ('low', float),
+    ('close', float),
+    ('volume', float),
+    ('bar_wap', float),  # will be zeroed by sampler if not filled
+]
+
+
+class KucoinMktPair(Struct, frozen=True):
+    '''
+    Kucoin's pair format
+
+    '''
+
+    baseCurrency: str
+    baseIncrement: float
+    baseMaxSize: float
+    baseMinSize: float
+    enableTrading: bool
+    feeCurrency: str
+    isMarginEnabled: bool
+    market: str
+    minFunds: float
+    name: str
+    priceIncrement: float
+    priceLimitRate: float
+    quoteCurrency: str
+    quoteIncrement: float
+    quoteMaxSize: float
+    quoteMinSize: float
+    symbol: str
+
+
+class AccountTrade(Struct, frozen=True):
+    '''
+    Historical trade format
+
+    '''
+
+    id: str
+    currency: str
+    amount: float
+    fee: float
+    balance: float
+    accountType: str
+    bizType: str
+    direction: Literal['in', 'out']
+    createdAt: float
+    context: list[str]
+
+
+class AccountResponse(Struct, frozen=True):
+    currentPage: int
+    pageSize: int
+    totalNum: int
+    totalPage: int
+    items: list[AccountTrade]
+
+
+class KucoinTrade(Struct, frozen=True):
+    '''
+    Real-time trade format
+
+    '''
+
+    bestAsk: float
+    bestAskSize: float
+    bestBid: float
+    bestBidSize: float
+    price: float
+    sequence: float
+    size: float
+    time: float
+
+
+class KucoinL2(Struct, frozen=True):
+    '''
+    Real-time L2 order book format
+
+    '''
+
+    asks: list[list[float]]
+    bids: list[list[float]]
+    timestamp: float
+
+
+class KucoinMsg(Struct, frozen=True):
+    type: str
+    topic: str
+    subject: str
+    data: list[KucoinTrade | KucoinL2]
+
+
+class BrokerConfig(Struct, frozen=True):
+    key_id: str
+    key_secret: str
+    key_passphrase: str
+
+
+def get_config() -> BrokerConfig | None:
+    conf, path = config.load()
+
+    section = conf.get('kucoin')
+
+    if section is None:
+        log.warning('No config section found for kucoin in config')
+        return None
+
+    return BrokerConfig(**section)
+
+
+class Client:
+    def __init__(self) -> None:
+        self._pairs: dict[str, KucoinMktPair] = {}
+        self._bars: list[list[float]] = []
+        self._key_id: str
```

Comment:
are these duplicate with the fields of `BrokerConfig`?
If so why do you need both sets?

Reply:
just do `self._config: BrokerConfig` no?

fwiw `msgspec.Struct`s don't do any validation, so packing it in one without doing anything else is basically a noop.